quickwit-storage/src/storage_with_upload_cache.rs

```diff
@@ -0,0 +1,885 @@
+            return Ok(());
+        }
+
+        if let Some(item_num_bytes) = locked_disk_cache.lru.pop(&path.to_path_buf()) {
+            if self.local_storage.delete(path).await.is_err() {
+                return Ok(());
+            }
+            locked_disk_cache.num_bytes -= item_num_bytes;
+            locked_disk_cache.num_files -= 1;
+        }
+
+        while self.exceeds_capacity(
+            locked_disk_cache.num_bytes + payload_length,
+            locked_disk_cache.num_files + 1,
+        ) {
+            if let Some((item_path, item_num_bytes)) = locked_disk_cache.lru.pop_lru() {
+                if self
+                    .local_storage
+                    .delete(item_path.as_path())
+                    .await
+                    .is_err()
+                {
+                    return Ok(());
+                }
+                locked_disk_cache.num_bytes -= item_num_bytes;
+                locked_disk_cache.num_files -= 1;
+            }
+        }
+
+        self.local_storage.put(path, payload).await?;
+        locked_disk_cache.num_bytes += payload_length;
+        locked_disk_cache.num_files += 1;
+        locked_disk_cache
+            .lru
+            .put(path.to_path_buf(), payload_length);
+        let _ = self.save_state(&locked_disk_cache).await;
+        Ok(())
+    }
+
+    async fn copy_to_file(&self, path: &Path, output_path: &Path) -> StorageResult<()> {
+        let mut locked_disk_cache = self.disk_cache.lock().await;
+        if locked_disk_cache.lru.get(&path.to_path_buf()).is_none() {
+            return self.remote_storage.copy_to_file(path, output_path).await;
+        }
+
+        let _ = self.save_state(&locked_disk_cache).await;
+        self.local_storage.copy_to_file(path, output_path).await
+    }
+
+    async fn get_slice(&self, path: &Path, range: Range<usize>) -> StorageResult<Bytes> {
+        let mut locked_disk_cache = self.disk_cache.lock().await;
+        if locked_disk_cache.lru.get(&path.to_path_buf()).is_none() {
+            return self.remote_storage.get_slice(path, range).await;
+        }
+
+        let _ = self.save_state(&locked_disk_cache).await;
+        self.local_storage.get_slice(path, range).await
+    }
+
+    async fn get_all(&self, path: &Path) -> StorageResult<Bytes> {
+        let mut locked_disk_cache = self.disk_cache.lock().await;
+        if locked_disk_cache.lru.get(&path.to_path_buf()).is_none() {
+            return self.remote_storage.get_all(path).await;
+        }
+
+        let _ = self.save_state(&locked_disk_cache).await;
+        self.local_storage.get_all(path).await
+    }
+
+    async fn delete(&self, path: &Path) -> StorageResult<()> {
+        self.remote_storage.delete(path).await?;
+
+        let mut locked_disk_cache = self.disk_cache.lock().await;
+        if let Some(item_num_bytes) = locked_disk_cache.lru.pop(&path.to_path_buf()) {
+            if self.local_storage.delete(path).await.is_err() {
```

Comment:
Can you check the error and log if it is not a file missing error?

Also, if the error is about the file missing locally, we should update the state.